Code/EnginePlugins/RecastPlugin/Components/DetourCrowdAgentComponent.cpp

```diff
@@ -0,0 +1,251 @@
+  s << m_fHeight;
+  s << m_fMaxSpeed;
+  s << m_fMaxAcceleration;
+  s << m_fStoppingDistance;
+}
+
+void ezDetourCrowdAgentComponent::DeserializeComponent(ezWorldReader& stream)
+{
+  SUPER::DeserializeComponent(stream);
+  ezStreamReader& s = stream.GetStream();
+
+  s >> m_fRadius;
+  s >> m_fHeight;
+  s >> m_fMaxSpeed;
+  s >> m_fMaxAcceleration;
+  s >> m_fStoppingDistance;
+}
+
+void ezDetourCrowdAgentComponent::FillAgentParams(ezDetourCrowdAgentParams& out_params) const
+{
+  out_params.m_fRadius = m_fRadius;
+  out_params.m_fHeight = m_fHeight;
+  out_params.m_fMaxSpeed = m_fMaxSpeed;
+  out_params.m_fMaxAcceleration = m_fMaxAcceleration;
+}
+
+void ezDetourCrowdAgentComponent::SetTargetPosition(const ezVec3& vPosition)
+{
+  m_vTargetPosition = vPosition;
+  m_PathToTargetState = ezAgentPathFindingState::HasTargetWaitingForPath;
+  m_bTargetDirty = true;
+}
+
+void ezDetourCrowdAgentComponent::ClearTargetPosition()
+{
+  m_PathToTargetState = ezAgentPathFindingState::HasNoTarget;
+}
+
+void ezDetourCrowdAgentComponent::OnActivated()
+{
+  SUPER::OnActivated();
+}
+
+void ezDetourCrowdAgentComponent::OnDeactivated()
+{
+  SUPER::OnDeactivated();
+}
+
+void ezDetourCrowdAgentComponent::OnSimulationStarted()
+{
+  SUPER::OnSimulationStarted();
+
+  ezCharacterControllerComponent* pCC = nullptr;
+  if (GetOwner()->TryGetComponentOfBaseType<ezCharacterControllerComponent>(pCC))
+  {
+    m_hCharacterController = pCC->GetHandle();
+  }
+}
+
+void ezDetourCrowdAgentComponent::SyncTransform(const ezVec3& vPosition, const ezVec3& vVelocity)
+{
+  m_vVelocity = m_vVelocity;
+
+  GetOwner()->SetGlobalPosition(vPosition);
+
+  //ezSimdTransform xform = GetOwner()->GetGlobalTransformSimd();
+  //xform.m_Position.Set(ezRcPos(pDtAgent->npos));
+  //xform.m_Rotation.
+  //GetOwner()->SetGlobalTransform(xform);
+}
+
+//////////////////////////////////////////////////////////////////////////
+//////////////////////////////////////////////////////////////////////////
+//////////////////////////////////////////////////////////////////////////
+//////////////////////////////////////////////////////////////////////////
+//////////////////////////////////////////////////////////////////////////
+
+ezDetourCrowdAgentComponentManager::ezDetourCrowdAgentComponentManager(ezWorld* pWorld)
+  : SUPER(pWorld)
+{
+}
+ezDetourCrowdAgentComponentManager::~ezDetourCrowdAgentComponentManager() = default;
+
+void ezDetourCrowdAgentComponentManager::Initialize()
+{
+  SUPER::Initialize();
+
+  m_pDetourCrowdModule = GetWorld()->GetOrCreateModule<ezDetourCrowdWorldModule>();
+
+  {
+    auto desc = EZ_CREATE_MODULE_UPDATE_FUNCTION_DESC(ezDetourCrowdAgentComponentManager::Update, this);
+    desc.m_Phase = ezWorldModule::UpdateFunctionDesc::Phase::PostAsync;
+    desc.m_bOnlyUpdateWhenSimulating = true;
+    desc.m_fPriority = 0.0f;
+
+    RegisterUpdateFunction(desc);
+  }
+}
+
+void ezDetourCrowdAgentComponentManager::Deinitialize()
+{
+  SUPER::Deinitialize();
+}
```

Comment:
redundant, unless you want to add something here later

Reply:
@jankrassnigg Thanks for the review! There are a few more things I want to do:
- The component should update game object rotation so it's always looks in the direction of horizontal velocity (can be turned on/off)
- Apply movement through `ezCharacterControllerComponent` if there is one
- Debug drawing
- Don't call `ezDetourCrowdAgentComponent::SyncTransform` if the agent is not moving
- Probably add some comments to `ezDetourCrowdAgentComponentManager::Update`
- More testing and bugfixing